docs/man/fsc.1
.\" ##########################################################################
.\" #                      __                                                #
.\" #      ________ ___   / /  ___     Scala 2 On-line Manual Pages          #
.\" #     / __/ __// _ | / /  / _ |    (c) 2002-2007, LAMP/EPFL              #
.\" #   __\ \/ /__/ __ |/ /__/ __ |                                          #
.\" #  /____/\___/_/ |_/____/_/ | |    http://scala-lang.org/                #
.\" #                           |/                                           #
.\" ##########################################################################
.\"
.\" Process this file with nroff -man scala.1
.\"
.TH fsc 1  "January 18, 2007" "version 0.4" "USER COMMANDS"
.\"
.\" ############################## NAME ###############################
.\"
.SH NAME
.PP
\fBfsc\fR \- Fast offline compiler for the Scala 2 language
.\"
.\" ############################## SYNOPSIS ###############################
.\"
.SH SYNOPSIS
.PP
\fBfsc\fR  [ <\fIoptions\fR> ] <\fIsource files\fR>
.\"
.\" ############################## PARAMETERS ###############################
.\"
.SH PARAMETERS
.PP
.TP
<\fIoptions\fR>
Command line options. See \fBOPTIONS\fR below.
.TP
<\fIsource files\fR>
One or more source files to be compiled (such as MyClass.scala).
.\"
.\" ############################## OPTIONS ###############################
.\"
.SH OPTIONS
.PP
The offline compiler supports all options of \fBscalac\fR plus the following:
.PP
.TP
\fB\-reset\fR 
Reset compile server caches.
.TP
\fB\-shutdown\fR 
Shut down the compilation daemon.  The daemon attempts to restart itself as necessary, but sometimes an explicit shutdown is required. A common example is if jars on the class path have changed.
.TP
\fB\-server\fR <\fIhostname:portnumber\fR> 
Specify compile server host at port number.  Usually this option is not needed.  Note that the hostname must be for a host that shares the same filesystem.
.TP
\fB\-J\fR <\fIflag\fR> 
Pass <flag> directly to the Java VM for the compilation daemon.
.\"
.\" ############################## DESCRIPTION ###############################
.\"
.SH DESCRIPTION
.PP
The \fBfsc\fR tool submits Scala compilation jobs to a compilation daemon. The first time it is executed, the daemon is started automatically. On subsequent runs, the same daemon can be reused, thus resulting in a faster compilation. The tool is especially effective when repeatedly compiling with the same class paths, because the compilation daemon can reuse a compiler instance.
.PP
The compilation daemon is smart enough to flush its cached compiler when the class path changes.  However, if the contents of the class path change, for example due to upgrading a library, then the daemon should be explicitly shut down with \fB-shutdown\fR.
.PP
Note that the \fBscala\fR script runner will also use the offline compiler by default, with the same advantages and caveats.
.\"
.\" ############################## EXAMPLE ###############################
.\"
.SH EXAMPLE
.PP
The following session shows a typical speed up due to using the offline compiler.

.nf
> fsc -verbose -d /tmp test.scala
...
[Port number: 32834]
[Starting new Scala compile server instance]
[Classpath = ...]
[loaded directory path ... in 692ms]
...
[parsing test.scala]
...
[total in 943ms]

> fsc -verbose -d /tmp test.scala
...
[Port number: 32834]
[parsing test.scala]
...
[total in 60ms]

> fsc -verbose -d /tmp test.scala
...
[Port number: 32834]
[parsing test.scala]
...
[total in 42ms]

> fsc -verbose -shutdown
[Scala compile server exited]

.fi
.\"
.\" ############################## ENVIRONMENT ###############################
.\"
.SH ENVIRONMENT
.PP
.TP
\fBJAVACMD\fR
Specify the \fBjava\fR command to be used for running the Scala code.  Arguments may be specified as part of the environment variable; spaces, quotation marks, etc., will be passed directly to the shell for expansion.
.TP
\fBJAVA_OPTS\fR
Specify the options to be passed to the \fBjava\fR command defined by \fBJAVACMD\fR.
.IP
With Java 1.5 (or newer) one may for example configure the memory usage of the JVM as follows: JAVA_OPTS="-Xmx512M -Xms16M -Xss16M"
.IP
With GNU Java one may configure the memory usage of the GIJ as follows: JAVA_OPTS="--mx512m --ms16m"
.IP

.\"
.\" ############################## EXIT STATUS ###############################
.\"
.SH "EXIT STATUS"
.PP
\fBfsc\fR returns a zero exit status if it succeeds to compile the specified input files. Non zero is returned in case of failure.
.\"
.\" ############################## AUTHOR ###############################
.\"
.SH AUTHOR
.PP
Written by Martin Odersky and other members of the Scala team.
.\"
.\" ############################## REPORTING BUGS ###############################
.\"
.SH "REPORTING BUGS"
.PP
Report bugs to <scala@listes.epfl.ch>.
.\"
.\" ############################## COPYRIGHT ###############################
.\"
.SH COPYRIGHT
.PP
This is open-source software, available to you under a BSD-like license. See accomponying "copyright" or "LICENSE" file for copying conditions. There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.\"
.\" ############################## SEE ALSO ###############################
.\"
.SH "SEE ALSO"
.PP
\fBsbaz\fR(1), \fBscala\fR(1), \fBscalac\fR(1), \fBscaladoc\fR(1), \fBscalap\fR(1)


debian/changelog
scala (2.6.1-1) unstable; urgency=low

  * New upstream release

 -- Lex Spoon <lex@debian.org>  Tue, 15 Jan 2008 18:08:31 +0100

scala (2.6.0-1) unstable; urgency=low

  * New upstream release

 -- Lex Spoon <lex@debian.org>  Fri, 14 Sep 2007 15:22:05 +0100

scala (2.5.0-1) unstable; urgency=low

  * New upstream release
  * Depend only on java-gcj-compat, not java-gcj-compat-dev

 -- Lex Spoon <lex@debian.org>  Tue, 22 May 2007 16:51:47 +0200

scala (2.4.0-1) unstable; urgency=low

  * New upstream version.

  * Deleting all mention of the Sun JVM, just because some
    DD's find it bothersome.

  * Deleted mention of JDK5-specific classes, while awaiting
    the Sun JVM to become open source.  (Closes: #409785)

  * Added java-gcj-compat-dev as an install dependency.  (Closes: #408353)

 -- Lex Spoon <lex@debian.org>  Sun, 25 Mar 2007 11:56:18 +0200

scala (2.3.2-1) unstable; urgency=low

  * New upstream version

 -- Lex Spoon <lex@debian.org>  Fri, 19 Jan 2007 14:45:51 +0100

scala (2.3.0-1) unstable; urgency=low

  * New upstream version
  * Added a Java compiler as a build-dependency

 -- Lex Spoon <lex@debian.org>  Fri, 24 Nov 2006 16:02:09 +0100

scala (2.2.0-1) unstable; urgency=low

  * New upstream release

 -- Lex Spoon <lex@debian.org>  Mon, 30 Oct 2006 21:35:10 -0500

scala (2.1.5-2) unstable; urgency=low

  * Rearranged to have three packages instead of four.

 -- Lex Spoon <lex@debian.org>  Wed, 28 Jun 2006 17:16:02 +0200

scala (2.1.5-1) unstable; urgency=low

  * upstream update.

 -- Lex Spoon <lex@debian.org>  Sun, 9 Jun 2006 14:45:26 +0100

scala (2.1.1-1) unstable; urgency=low

  * Initial Release.

 -- Lex Spoon <lex@debian.org>  Sun, 26 Mar 2006 14:45:26 +0100



src/manual/scala/man1/fsc.scala
/* NSC -- new Scala compiler
 * Copyright 2005-2007 LAMP/EPFL
 * @author Stephane Micheloud
 */
//$Id$

package scala.man1

object fsc extends Command {
  import _root_.scala.tools.docutil.ManPage._

  protected def cn = new Error().getStackTrace()(0).getClassName()
  override def lastModified = "January 18, 2007"

  val name = Section("NAME",

    MBold(command) & " " & NDash & " Fast offline compiler for the " &
    Link("Scala 2", "http://scala-lang.org/") & " language")

  val synopsis = Section("SYNOPSIS",

    CmdLine(" [ " & Argument("options") & " ] " &
            Argument("source files")))

  val parameters = scalac.parameters

  val description = Section("DESCRIPTION",

    "The "&MBold("fsc")&" tool submits Scala compilation jobs to " &
    "a compilation daemon. "&
    "The first time it is executed, the daemon is started automatically. "&
    "On subsequent "&
    "runs, the same daemon can be reused, thus resulting in a faster compilation. "&
    "The tool is especially effective when repeatedly compiling with the same "&
    "class paths, because the compilation daemon can reuse a compiler instance.",

    "The compilation daemon is smart enough to flush its cached compiler "&
    "when the class path changes.  However, if the contents of the class path "&
    "change, for example due to upgrading a library, then the daemon "&
    "should be explicitly shut down with " & MBold("-shutdown") & ".",

    "Note that the " & Link(MBold("scala"), "scala.html") & " script runner " &
    "will also use " &
    "the offline compiler by default, with the same advantages and caveats.")

  val options = Section("OPTIONS",

      "The offline compiler supports " &
      Link("all options of " & MBold("scalac"), "scalac.html#options") &
      " plus the following:",

      DefinitionList(
        Definition(
          CmdOption("reset"),
          "Reset compile server caches."),
        Definition(
          CmdOption("shutdown"),
          "Shut down the compilation daemon.  The daemon attempts to restart "&
          "itself as necessary, but sometimes an explicit shutdown is required. "&
          "A common example is if jars on the class path have changed."),
        Definition(
          CmdOption("server", Argument("hostname:portnumber")),
          "Specify compile server host at port number.  Usually this option " &
          "is not needed.  Note that the hostname must be for a host that shares " &
          "the same filesystem."),
        Definition(
          CmdOption("J", Argument("flag")),
          "Pass <flag> directly to the Java VM for the compilation daemon.")
    ))

  val example = Section("EXAMPLE",

      "The following session shows a typical speed up due to using the "&
      "offline compiler.",

      CodeSample(
      """> fsc -verbose -d /tmp test.scala
        |...
        |[Port number: 32834]
        |[Starting new Scala compile server instance]
        |[Classpath = ...]
        |[loaded directory path ... in 692ms]
        |...
        |[parsing test.scala]
        |...
        |[total in 943ms]
        |
        |> fsc -verbose -d /tmp test.scala
        |...
        |[Port number: 32834]
        |[parsing test.scala]
        |...
        |[total in 60ms]
        |
        |> fsc -verbose -d /tmp test.scala
        |...
        |[Port number: 32834]
        |[parsing test.scala]
        |...
        |[total in 42ms]
        |
        |> fsc -verbose -shutdown
        |[Scala compile server exited]
        |""".stripMargin))

  val environment = Section("ENVIRONMENT",

    DefinitionList(
      Definition(
        MBold("JAVACMD"),
        "Specify the " & MBold("java") & " command to be used " &
        "for running the Scala code.  Arguments may be specified " &
        "as part of the environment variable; spaces, quotation marks, " &
        "etc., will be passed directly to the shell for expansion."),
      Definition(
        MBold("JAVA_OPTS"),
        SeqPara(
          "Specify the options to be passed to the " & MBold("java") &
          " command defined by " & MBold("JAVACMD") & ".",

          "With Java 1.5 (or newer) one may for example configure the " &
          "memory usage of the JVM as follows: " &
          Mono("JAVA_OPTS=\"-Xmx512M -Xms16M -Xss16M\""),

          "With " & Link("GNU Java", "http://gcc.gnu.org/java/") & " one " &
          "may configure the memory usage of the GIJ as follows: " &
          Mono("JAVA_OPTS=\"--mx512m --ms16m\"")
        ))))

  val exitStatus = Section("EXIT STATUS",

    MBold(command) & " returns a zero exit status if it succeeds to " &
    "compile the specified input files. Non zero is returned in case " &
    "of failure.")

  val seeAlso = Section("SEE ALSO",

    Link(Bold("sbaz") & "(1)", "sbaz.html") & ", " &
    Link(Bold("scala") & "(1)", "scala.html") & ", " &
    Link(Bold("scalac") & "(1)", "scalac.html") & ", " &
    Link(Bold("scaladoc") & "(1)", "scaladoc.html") & ", " &
    Link(Bold("scalap") & "(1)", "scalap.html"))

  def manpage = new Document {
    title = command
    date = lastModified
    author = "Lex Spoon"
    version = "0.4"
    sections = List(
      name,
      synopsis,
      parameters,
      options,
      description,
      example,
      environment,
      exitStatus,
      authors,
      bugs,
      copyright,
      seeAlso)
  }
}
